arch/arm/src/xmc4/xmc4_tickless.c

```diff
@@ -0,0 +1,576 @@
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void xmc4_timing_handler(void)
+{
+  tmrinfo("Overflow");
+  g_tickless.overflow++;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_timer_initialize
+ *
+ * Description:
+ *   Initializes all platform-specific timer facilities.  This function is
+ *   called early in the initialization sequence by up_initialize().
+ *   On return, the current up-time should be available from
+ *   up_timer_gettime() and the interval timer is ready for use (but not
+ *   actively timing.
+ *
+ *   Provided by platform-specific code and called from the architecture-
+ *   specific logic.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Called early in the initialization sequence before any special
+ *   concurrency protections are required.
+ *
+ ****************************************************************************/
+
+void up_timer_initialize(void)
+{
+#ifdef CONFIG_SCHED_TICKLESS_LIMIT_MAX_SLEEP
+  uint64_t max_delay;
+#endif
+  int ret;
+
+  g_tickless.pending = false;
+  g_tickless.overflow = 0;
+  g_tickless.frequency = USEC_PER_SEC / (uint32_t)CONFIG_USEC_PER_TICK;
+
+  /* Enable CCU clock */
+
+  putreg32(SCU_CLK_CCUC, XMC4_SCU_CLKSET);
+
+  /* Enable CCU clock during sleep */
+
+  putreg32(SCU_SLEEPCR_CCUCR | SCU_SLEEPCR_SYSSEL, XMC4_SCU_SLEEPCR);
+
+  uint32_t divisor;
+  uint32_t pssiv;
+  ret = xmc4_ccu4_divisor(USEC_PER_SEC / (uint32_t)CONFIG_USEC_PER_TICK,
+                          &divisor,
+                          &pssiv);
+  g_tickless.frequency = BOARD_CCU_FREQUENCY / divisor;
+  if (ret < 0)
+    {
+      tmrerr("ERROR: xmc4_ccu4_divisor failed: %d\n", ret);
+      return ret;
+    }
+
+  tmrinfo("frequency=%lu, divisor=%lu, cmr=%08lx\n",
+          (unsigned long)g_tickless.frequency, (unsigned long)divisor,
+          (unsigned long)pssiv);
+
+  /* Initialize Interval Timer
+   *
+   * Ths timer is configured to be a oneshot timer, that has
+   * a resolution that matches the USEC_PER_TICK, and
+   * will be started in up_timer_start and uses its period
+   * (not compare value) to trigger an interrupt.
+   */
+
+  /* Apply reset */
+
+  putreg32(SCU_PR0_CCU41RS, XMC4_SCU_PRSET0);
+  putreg32(SCU_PR0_CCU41RS, XMC4_SCU_PRCLR0);
+
+  /* Enable CC40 Slice */
+
+  putreg32(CCU4_GIDLC_CS0I_MASK, XMC4_CCU41_GIDLC);
+
+  /* Enable the prescaler and set value */
+
+  putreg32(CCU4_GIDLC_SPRB_MASK, XMC4_CCU41_GIDLC);
+  putreg32(pssiv, XMC4_CCU41_CC40PSC);
+
+#ifdef CONFIG_SCHED_TICKLESS_LIMIT_MAX_SLEEP
+  max_delay = (float)UINT16_MAX * USEC_PER_SEC / g_tickless.frequency;
```

Comment:
Maybe `float` can be avoided?

Reply:
nop, it can't ;)

Reply:
why need float if the final result is an integer.